### PR TITLE
[MRG] Configurable subsequent DTW path computation

### PR DESCRIPTION
**Pull request by AKuederle**
See #137

### CODE REVIEW

File: tslearn/tests/test_metrics.py [hunk 1]
```diff
@@ -42,6 +43,24 @@ def test_dtw_subseq():
...
+    subseq = to_time_series(subseq)
+    longseq = to_time_series(longseq)
+    cost_matrix = tslearn.metrics.njit_accumulated_matrix_subsequence(subseq, longseq)
+
+    path = tslearn.metrics.return_path_subsequence(cost_matrix, 3)
+    np.testing.assert_equal(path, [(0, 2), (1, 3)])
+
+    path = tslearn.metrics.return_path_subsequence(cost_matrix, 1)
+    np.testing.assert_equal(path, [(0, 0), (1, 1)])
+
```
Comment (rtavenar) [inline tslearn/tests/test_metrics.py]: Maybe adding a test that asserts that on some randomly-generated data, `dtw_subsequence_path` gives the same path / cost as `return_path_subsequence` when starting from the argmin?

**Comment by rtavenar** [comment]
As stated in #137, this use case should be properly documented as I am unsure how straight forward it can be for an end user to understand it properly. I plan to improve on the current example for DTW path soon, maybe both notebooks could share a more or less common appearance, something like [this example](https://tslearn.readthedocs.io/en/latest/_images/dtw.png)?

Also, I am unsure of why a Travis job was failing: I restarted the corresponding job to make sure this is reproducible (it was an apparently unrelated test that was failing, so...)

Finally, this PR will change the public API for `tslearn` (without regression, though), so it should be shipped with `tslearn` 0.3. I will prepare a `dev` branch asap for all the improvements to be included in `tslearn` 0.3, and this PR should be merged to that `dev` branch instead of the `master` branch then.

**Comment by AKuederle** [comment]
Thank your for your comment! I will look into it and update the pull request!

**Comment by rtavenar** [comment]
OK, I prepared a new DTW visualization example for the docs in `dev` branch, cf [here](https://github.com/rtavenar/tslearn/blob/dev/tslearn/docs/examples/plot_dtw.py).

Maybe you could use more or less the same kind of visualization but:
* Use `tslearn` function to compute cross-similarity matrix (and update my example so that it does the same, since the function becomes public with your PR) instead of `cdist` from `scipy`;
* have several paths exhibited for several end points.

What do you think?

**Comment by AKuederle** [comment]
Sounds good! Btw. should we drop the njit prefix for the functions, which will become public API. I guess the naming was introduced, because your transitioned to numba from Cython, but for public API, this might be confusing.

**Comment by rtavenar** [comment]
Yep! Good idea!

File: tslearn/metrics.py [hunk 2]
```diff
@@ -400,7 +433,7 @@ def dtw_subsequence_path(subseq, longseq):
 
     Compared to traditional DTW, here, border constraints on admissible paths
     :math:`P` are relaxed such that :math:`P_0 = (0, ?)` and
-    :math:`P_L = (N-1, ?)` where :math:`L` is the length of the considered path
+    :math:`P_0 = (0, ?)` where :math:`L` is the length of the considered path
```
Comment (rtavenar) [inline tslearn/metrics.py]: Hum, I don't agree with your edit. I think the original formulation was the correct one.

File: tslearn/metrics.py [hunk 3]
```diff
@@ -362,10 +361,44 @@ def njit_accumulated_matrix_subsequence(subseq, longseq):
 
 
 @njit()
-def _return_path_subsequence(acc_cost_mat):
+def calculate_subsequence_path(acc_cost_mat, idx_path_end):
```
Comment (rtavenar) [inline tslearn/metrics.py]: `calculate_subsequence_path` is a bit of a complicated name for public API no? What about `subsequence_path`?

**Comment by rtavenar** [comment]
> I added the example and a little bit of explanation. Do you think this is sufficient to understand the usecase?

Yep, I think it helps a lot!

File: tslearn/docs/examples/plot_sdtw.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+"""
+sDTW multi path matching
+========================
+
+This example illustrates how subsequent DTW can be used to find multiple
+matches of a sequence in a longer sequence.
+
+A potential usecase is to identify the occurrence of certain events in
+continuous sensor signals. As one example Barth et al. [1] used this method
+to find stride in sensor recordings of gait
```
Comment (rtavenar) [inline tslearn/docs/examples/plot_sdtw.py]: Add "." at the end of the sentence.

File: tslearn/docs/examples/plot_sdtw.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+The example demonstrates the use of the functions
+`accumulated_subsequence_cost_matrix` and `calculate_subsequence_path`
+to manually calculate warping paths from multiple potential alignments. If
+you are only interested in finding the optimal alignment, you can directly use
+`dtw_subsequence_path`.
+
+[1] Barth, et al. (2013). Subsequence dynamic time warping as a method for
```
Comment (rtavenar) [inline tslearn/docs/examples/plot_sdtw.py]: Looks like the reference here does not render nicely. From `sklearn` docs (eg [here](https://scikit-learn.org/stable/auto_examples/applications/plot_species_distribution_modeling.html#sphx-glr-auto-examples-applications-plot-species-distribution-modeling-py), it seems that using bullet lists for references is more convenient.

File: tslearn/metrics.py [hunk 6]
```diff
@@ -362,10 +361,44 @@ def njit_accumulated_matrix_subsequence(subseq, longseq):
...
+    idx_path_end: int
+        The end position of the matched subsequence in the longer sequence.
+
+    Returns
+    -------
+    path: list of tuples of integer pairs
+        Matching path represented as a list of index pairs. In each pair, the
+        first index corresponds to `subseq` and the second one corresponds to
+        `longseq`. The startpoint of the Path is :math:`P_0 = (0, ?)` and it
+        ends at :math:`P_0 = (0, idx_path_end)`
```
Comment (rtavenar) [inline tslearn/metrics.py]: Again, I disagree (cf comment above): the last element in the path should be of the form `(sz1-1, ?)`

**Comment by rtavenar** [comment]
Also you should add `accumulated_subsequence_cost_matrix` and `calculate_subsequence_path` to the list of documented items in metrics module, which is defined in `tslearn/docs/gen_modules/tslearn.metrics.rst`

File: tslearn/docs/examples/plot_sdtw.py (hunk 5 above)
Comment (AKuederle) [reply tslearn/docs/examples/plot_sdtw.py]: This does not seem to be consistent. At other instances they used the same citation style I used (e.g. [here](https://github.com/scikit-learn/scikit-learn/blob/master/examples/covariance/plot_mahalanobis_distances.py)

File: tslearn/docs/examples/plot_sdtw.py (hunk 4 above)
Comment (AKuederle) [reply tslearn/docs/examples/plot_sdtw.py]: Done

File: tslearn/metrics.py (hunk 3 above)
Comment (AKuederle) [reply tslearn/metrics.py]: I renamed them to `subsequence_path` and `subsequence_cost_matrix`

**Comment by AKuederle** [comment]
> Also you should add `accumulated_subsequence_cost_matrix` and `calculate_subsequence_path` to the list of documented items in metrics module, which is defined in `tslearn/docs/gen_modules/tslearn.metrics.rst`

Done

File: tslearn/docs/examples/plot_sdtw.py (hunk 5 above)
Comment (rtavenar) [reply tslearn/docs/examples/plot_sdtw.py]: You are perfectly right. My main concern in fact was about the obtained formatting: in the case you mention, the first example is not properly rendered (the name of the journal starts in bold on the first line and then continues without bold on the second line) while formatting of the second item is OK (title + authors in bold, the rest in non-bold). If you can find a way to have some reasonable rendering in your case that would be perfect.

Comment (AKuederle) [reply tslearn/docs/examples/plot_sdtw.py]: I will have a look and make sure it looks good in the hml output :)

Comment (rtavenar) [reply tslearn/docs/examples/plot_sdtw.py]: OK, great! If you build the doc locally, you may have to uncomment the following line (at least I have to on my side :)

https://github.com/rtavenar/tslearn/blob/master/tslearn/docs/conf.py#L23

File: tslearn/metrics.py (hunk 2 above)
Comment (AKuederle) [reply tslearn/metrics.py]: I don't know why I changed that. You are absolutely right

**Comment by rtavenar** [comment]
Does the doc compilation work on your side? On my side, I get the following error:

```
WARNING: autodoc: failed to import function 'subsequence_cost_matrix' from module 'metrics'; the following exception was raised:
No module named 'metrics'
WARNING: autodoc: failed to import function 'subsequence_path' from module 'metrics'; the following exception was raised:
No module named 'metrics'
```

Not sure why I get those, maybe it is just a local config issue, though.

**Comment by AKuederle** [comment]
> Does the doc compilation work on your side? On my side, I get the following error:
> 
> ```
> WARNING: autodoc: failed to import function 'subsequence_cost_matrix' from module 'metrics'; the following exception was raised:
> No module named 'metrics'
> WARNING: autodoc: failed to import function 'subsequence_path' from module 'metrics'; the following exception was raised:
> No module named 'metrics'
> ```
> 
> Not sure why I get those, maybe it is just a local config issue, though.

Mmh I got the same thing. I thought it was an issue with my local config. I couldn't figure out what the issue was

**Comment by AKuederle** [comment]
> > Does the doc compilation work on your side? On my side, I get the following error:
> > ```
> > WARNING: autodoc: failed to import function 'subsequence_cost_matrix' from module 'metrics'; the following exception was raised:
> > No module named 'metrics'
> > WARNING: autodoc: failed to import function 'subsequence_path' from module 'metrics'; the following exception was raised:
> > No module named 'metrics'
> > ```
> > 
> > 
> > Not sure why I get those, maybe it is just a local config issue, though.
> 
> Mmh I got the same thing. I thought it was an issue with my local config. I couldn't figure out what the issue was

I think I figured it out. The issue appears to be the `njit` decorator. With the decorator apidoc does not seem to be able to find the function.

**Comment by AKuederle** [comment]
A possible workaround would be to still have the njit funcs as private api and have a wrapper for as puplic api with documentation

**Comment by rtavenar** [comment]
Thanks a lot @AKuederle ! This is very good job!

**Pull request merged**